Add some preliminary guesser docs

lilyminium opened the pull request:
Supersedes #213, just a preliminary first attempt at jotting down what I think people would want to know and some possible gotchas with the new guessing interface. I've mostly tried to avoid duplicating the API docs where possible, except for the attribute guessing where there is some additional info in the user guide (which we can move over to core API docs if it belongs better there).

 If there are any other items that should really be added, please suggest them here!

<!-- readthedocs-preview mdanalysisuserguide start -->
----
📚 Documentation preview 📚: https://mdanalysisuserguide--409.org.readthedocs.build/en/409/

<!-- readthedocs-preview mdanalysisuserguide end -->

lilyminium committed:
add some preliminary guesser docs

lilyminium committed:
fix ipython blocks

RMeli commented on doc/source/formats/guessers/default.rst:
```diff
@@ -0,0 +1,109 @@
+
+.. warning::
+
+    The default guesser has been created to reproduce pre-v2.8.0 MDAnalysis guessing behaviour as much as possible. **Default behaviours will change in MDAnalysis version 3**, as detailed in deprecation warnings.
```
Do we have/want to include more details about what "as close as possible" means (i.e. are there small differences and corner cases)?

lilyminium replied:
As far as I recall the only real change is the ITPParser now has an extra Elements attribute, which is a bit specific to mention -- I've added a link to the 2.8.0 changelog for interested readers.

RMeli commented on doc/source/formats/guessers/default.rst:
```diff
@@ -0,0 +1,109 @@
+
+The default guesser guesses atom ``element``\ s and ``type``\ s using the same pathway; when atom ``type``\ s are guessed, they represent the atom ``element``. Atom elements are guessed from the atom name. The default guesser follows biological naming conventions, where atoms named "CA" are much more likely to represent an alpha-carbon than a calcium atom. This guesser is still relatively fragile ...
+
+The :meth:`~MDAnalysis.guesser.default_guesser.DefaultGuesser.guess_atom_element` method is used to guess atom elements or types following a process by which numbers, symbols, and some letters are stripped from the atom name and checked against a look-up table, as detailed in the linked API documentation. With this method, for example, "AO5*" would be guessed as "O", and "3hg2" as "H".
```
Worth linking the API documentation again, instead of mentioning the link above?

orbeckst replied:
done

RMeli commented on doc/source/formats/guessers/default.rst:
```diff
@@ -0,0 +1,109 @@
+Aromaticities
+-------------
+
+These are guessed using the :ref:`RDKit <RDKit-format>` converter by using the ``GetIsAromatic`` method.
```
I'm not sure about the implementation details here, but i assume RDKit remains an optional dependency. If this is the case, would it be worth adding that RDKit needs to be installed?

RMeli commented on doc/source/formats/guessers/default.rst:
```diff
@@ -0,0 +1,109 @@
+
+    d < f \cdot (R_1 + R_2)
+
+of each other, where :math:`R_1` and :math:`R_2` are the VdW radii
```
```suggestion
of each other, where :math:`R_1` and :math:`R_2` are the Van deer Waals radii
```

orbeckst replied:
https://en.wikipedia.org/wiki/Johannes_Diderik_van_der_Waals
```suggestion
of each other, where :math:`R_1` and :math:`R_2` are the van der Waals radii
```

RMeli commented on doc/source/formats/guessers/default.rst:
```diff
@@ -0,0 +1,109 @@
+    d < f \cdot (R_1 + R_2)
+
+of each other, where :math:`R_1` and :math:`R_2` are the VdW radii
+of the atoms and :math:`f` is an ad-hoc *fudge_factor*. This is
```
```suggestion
of the atoms and :math:`f` is an ad-hoc *fudge factor*. This is
```
Or use code-style formatting?

RMeli commented on doc/source/guessing.rst:
```diff
@@ -0,0 +1,153 @@
+  u.atoms.bonds
+
+
+In general, guessing at Universe creation works very similarly to guessing using the guess_TopologyAttrs interface documented below. The main difference is that passing guesser-specific keyword arguments such as ``fudge_factor`` and ``vdwradii`` into Universe creation is **now deprecated and will be removed in version 3.0**. Instead, we recommend specifying these arguments through an explicit cal ...
```
The first guess_TopologyAttrs should be formatted or be a reference (or both)?

lilyminium replied:
Done

RMeli commented on doc/source/guessing.rst:
```diff
@@ -0,0 +1,153 @@
+
+.. _guess-topologyAttrs:
+
+Guessing using the guess_TopologyAttrs interface
```
Format guess_TopologyAttributes?

RMeli commented on doc/source/guessing.rst:
```diff
@@ -0,0 +1,153 @@
+  u.atoms.types
+
+
+The context can be specified either using a string (e.g. ``"default"``) or an already created Guesser object. It may be convenient to pass in an already-created Guesser object if there are particular keywords you want to use in guessing methods, such as the ``fudge_factor``, ``vdwradii`` or ``lower_bound`` keywords for controlling bond guessing. However, if additional keyword arguments are passed ...
```
Format/reference to Guesser object?

lilyminium replied:
Done

RMeli commented on doc/source/guessing.rst:
```diff
@@ -0,0 +1,153 @@
+  print(len(u.bonds))
+
+
+However, the number of bonds doesn't change when the bonds are guessed again with a stricter criteria -- no new bonds are found.
```
Worth also being explicit about the fact that bonds not corresponding to the criteria are not removed (if I understand correctly)?

RMeli replied:
This is made super clear below, but might be good to foreshadow.

RMeli reviewed (commented):
Thanks @lilyminium for adding this documentation. I left some comments, but generally LGTM and it's very informative.

Do I understand correctly that there are currently no guessers other than the default?

Apologies if I din't provide some suggested changes for formatting, but I reviewed on mobile and back quotes are not an easy thing to add.

orbeckst commented on doc/source/formats/guessers/default.rst:
```diff
@@ -0,0 +1,109 @@
+-------------
+
+These are guessed using the :ref:`RDKit <RDKit-format>` converter by using the ``GetIsAromatic`` method.
+
```
```suggestion

.. note:: 
   RDKit needs to have been installed for aromaticity guessing to be available.
   RDKit is always installed when MDAnalysis was installed with conda-forge packages
   but this may not be the case when using other installation paths.
```

orbeckst reviewed (approved):
This looks good to me.

I made some suggestions (mostly in response to @RMeli ). I suggest to merge it, as least as the first pass for the guessers, and get it in. Improvements could be made later.

orbeckst committed:
Apply suggestions from code review

Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

orbeckst committed:
link to Guesser API

orbeckst committed:
Update guessing.rst

- address @RMeli comments
- minor grammar fixes
- moved note on radii guessing to explicit rst note

orbeckst commented:
@RMeli @lilyminium I made some edits, hopefully addressing most of @RMeli 's comments. Are we ok with merging so that we can make progress on the User Guide?

orbeckst commented on doc/source/formats/guessers/default.rst:
```diff
@@ -0,0 +1,115 @@
+
+These are guessed using the :ref:`RDKit <RDKit-format>` converter by using the ``GetIsAromatic`` method.
+
+.. note:: 
```
satisfy pre-commit hook check
```suggestion
.. note::
```

orbeckst commented on doc/source/guessing.rst:
```diff
@@ -0,0 +1,157 @@
+
+The order of the attributes guessed can matter in some cases. For example, bond guessing with the :class:`~MDAnalysis.guesser.default_guesser.DefaultGuesser` relies on looking up the vdW radii of the atoms involved by their atom ``type``. That means that for file formats where the atom ``type`` is not a valid element, the atom ``type`` must be forcefully re-guessed for bond-guessing to work.
+
+.. note:: 
```
satisy pre-commit hook check
```suggestion
.. note::
```

orbeckst reviewed (approved):
fix whitespace issues

orbeckst committed:
Apply suggestions from code review

fix whitespace to make pre-hook check happy

lilyminium committed:
fix ipython block

lilyminium committed:
add changes from review

lilyminium committed:
add link to changelog

lilyminium commented:
Thanks @RMeli and @orbeckst for the review and suggested changes! Correct, at the moment the default guesser is the only one that's implemented. I noticed an ipython block wasn't rendering correctly so will double check after CI runs and then would be happy to merge then.

RMeli commented:
Thanks. Ping me if you need me to press the merge button. ;)

lilyminium commented:
Huzzah it renders -- merging now!
